CORE-1249: support obtaining stat information.

slr71 opened the pull request:
This PR includes several changes to support obtaining stat information from this library. While I was modifying this library, I also took some time to update the dependencies and add some functions to make it easier to test this library from the REPL.

slr71 committed:
CORE-1249: added a function to extract multiple values from a cache

slr71 committed:
CORE-1249: updated some dependencies

slr71 committed:
CORE-1249: working but incomplete implementations of icat/paths-for-uuids and jargon/get-paths

slr71 committed:
CORE-1249: added .eastwood to .gitignore

slr71 committed:
CORE-1249: updated the multi-value cache code to store delays in the cache before waiting for the agent to complete

slr71 committed:
CORE-1249: abandoned the `action-runner` idea

ianmcorvidae commented:
I've been neglecting this, sorry about that. I'll try to give it a proper review later today, though of course it's still marked WIP and all that.

ianmcorvidae commented on src/clj_irods/cache_tools.clj:
```diff
@@ -44,9 +49,9 @@
   object with the error at a known key."
-  [action]
+  [action & args]
   (try+
-    (action)
+    (apply action args)
```
:+1:

ianmcorvidae commented on src/clj_irods/cache_tools.clj:
```diff
@@ -29,10 +29,15 @@
+            [clojure.set :as set]
             [clojure.tools.logging :as log]))
 
+(defmacro otel-with-subspan [[s] & body]
+  `(with-open [_# (otel/span-scope ~s)]
+     ~@body))
```
:+1:

ianmcorvidae commented on src/clj_irods/cache_tools.clj:
```diff
@@ -101,6 +106,33 @@
+      (otel/with-span [s ["agent for retrieving multiple values"]]
+        (let [ag  (agent nil)]
+          (send-via pool ag (fn [_nil] (otel-with-subspan [s] (do-or-error action uncached-ids))))
+          (delay (deref-vals (merge cached (store-multi cache location-fn ids ag))))))
```
I think the last line of this won't do quite what you want. I see a commit talking about adding delays to the cache before the agent finishes calculation, which this will do in some cases, but since the call to `store-multi` is inside the `delay` that gets returned, the delays won't make it into the cache until the `cached-or-retrieved-values` delay is dereferenced. I think this might work:

```suggestion
      (otel/with-span [s ["agent for retrieving multiple values"]]
        (let [ag (agent nil)]
          (send-via pool ag (fn [_nil] (otel-with-subspan [s] (do-or-error action uncached-ids))))
          (let [stored-delays (store-multi cache location-fn ids ag)]
            (delay (deref-vals (merge cached stored-delays)))))
```

(give or take any parens I did wrong, anyway)

slr71 replied:
Thanks for catching that.

slr71 commented:
No worries at all. I haven't completely finished implementing this, and I haven't had time to get back to it for a while.

ianmcorvidae commented on src/clj_irods/core.clj:
```diff
@@ -343,3 +343,9 @@
            (when (:has-jargon irods)
              (jargon/get-path irods uuid)))) uuid])))
+
+(defn uuids->paths
```
(I'm guessing this is part of the WIP aspect, leaving a note to myself to come back later)

ianmcorvidae reviewed (commented):
Looks generally good, I think one bit of the cache code isn't quite right, left a comment. Looks like the `clj-irods.core` stuff is what's outstanding, which I imagine will mostly just be implementing the `uuids->paths` function. (If I'm not mistaken, `uuid->path` will already use these values if they're cached?)

slr71 committed:
CORE-1249: Update src/clj_irods/cache_tools.clj

Co-authored-by: Ian McEwen <ianmcorvidae@ianmcorvidae.net>

slr71 committed:
CORE-1249: implemented uuids->paths

slr71 commented on src/clj_irods/core.clj:
```diff
@@ -83,7 +83,7 @@
            jargon-opts# (or (:jargon-opts ~cfg) {})
            use-jargon# (boolean jargon-cfg#)
-           use-icat# (have-icat)
+           use-icat# (and (:use-icat ~cfg true) (have-icat))
```
I wanted a way to easily disable direct ICAT access while I was testing.

slr71 commented:
I finally got around to finishing this up.

slr71 commented:
There's another tangentially related change that I'm going to include in this PR. Marking this as a draft PR again for now.

slr71 committed:
CORE-1249: made a slight change to clj-irods.validate/validate to make it easier to add conditional validations

slr71 committed:
CORE-1249: added a validation to check for the existence of entities associated with a collection of UUIDs

slr71 committed:
CORE-1249: updated clj-irods.validate/validate to include all failed values when multiple values are being validated

slr71 committed:
CORE-1249: fixed a bug that prevented UUIDs (rather than their string representations) from working with uuids->paths

slr71 committed:
CORE-1249: added a spot for REPL testing code that we don't have to worry about being checked in

slr71 committed:
CORE-1249: added a common namespace for functions to help with testing in the REPL

slr71 committed:
CORE-1249: added instructions for using clj-irods.repl-utils to the README

slr71 committed:
CORE-1249: updated some dependencies:

- org.clojure/clojure 1.10.2 -> 1.10.3: I didn't see any breaking changes mentioned in the change log
- org.cyverse/otel 0.2.0 -> 0.2.3: there aren't any breaking changes in this update
- org.cyverse/clojure-commons 2.8.3 -> 3.0.6: there aren't any breaking changes in this update that should affect us

slr71 committed:
CORE-1249: initial implementation of `stat` the `:md5` checksum still needs to be included for listings from the ICAT

slr71 committed:
CORE-1249: implmented `stat`

slr71 committed:
CORE-1249: added clj-irods.core/list-user-permissions and associated functions

slr71 committed:
CORE-1249: added functions for counting the number of files and folders in a folder

slr71 committed:
CORE-1249: added :file-size to stat objects obtained from ICAT listings, and fixed a bug in the epoch millis timestamp conversion

slr71 committed:
CORE-1249: fixed a bug in the multi-cache functions that caused problems when dealing with a partially cached result set

slr71 committed:
CORE-1249: omit the file size and MD5 checksum from stat information obtained from listings

slr71 commented:
This change should be ready for review again.

ianmcorvidae commented on src/clj_irods/core.clj:
```diff
@@ -343,3 +380,67 @@
+  "The paths associated with multiple uuids, returned as a map from uuid to
+  the corresponding path or nil"
+  [irods uuids]
+  (otel/with-span [s ["uuids->paths"]]
+    (cached-or-get irods
+      [(fn [cache? irods uuids]
+         (when (:has-icat irods)
+           (icat/paths-for-uuids irods uuids))) uuids]
+      [(fn [cache? irods uuids]
+         (when (:has-jargon irods)
+           (jargon/get-paths irods uuids))) uuids])))
```
I'd expect these to do something with the `cache?` variable. In particular, as currently written, this doesn't really use the functionality of `cached-or-get` much at all. If there's a cached jargon value but the icat is enabled, it will re-fetch the paths from the icat, since there's no guard on the first handler function only making real (non-cache) calls when `cache?` is true.

I'm not completely sure the best way to handle this case with these multi-fetch functions, though. I suppose ideally we'd allow each handler to return partial results, but maybe the effect would be mostly identical if it returns a cached value if and only if it can completely resolve it.

slr71 replied:
I struggled with this quite a bit when I first started working on this, and I admittedly punted at the time. There's a bit of a mismatch between the multi-fetch functions and the single fetch functions that I wasn't quite sure how to resolve at first. Currently, the check for cached items happens in `cached-or-retrieved-values`, which is probably where most of the mismatch is coming from. Now that I've had some more time to wrap my head around this, I think we could probably resolve it by writing a multi-fetch version of `cached-or-get` that tries function calls in the same order on a potentially progressively smaller set of unresolved IDs. Then that function can return either when the set of unresolved IDs is empty or all of the lookup methods have been attempted.

That's a big enough change that I think I'd like to dedicate a separate pull request to it. That will give also give me some time to work on other tasks for a while without feeling like this one is looming. I'll add a comment and create a Jira issue for this.

Thanks for mentioning this. I meant to come back to it and forgot.

slr71 replied:
I added the comment and created the Jira issue. Hopefully I can resume work on this in a couple of weeks.

ianmcorvidae reviewed (approved):
I think this looks reasonable. One comment that might merit changing, but it's probably fine regardless.

slr71 committed:
CORE-1249: added a comment as a reminder to further optimize the multi-fetch feature

slr71 commented:
Thanks for the review! I'm going to merge this now.
